Sources/TuistCache/Extensions/GraphTraverser+Extras.swift

```diff
@@ -0,0 +1,36 @@
+import TuistCore
+import TuistGraph
+
+extension GraphTraversing {
+    /// Returns the included based on the parameters.
+    public func filterIncludedTargets<GraphTargets: Collection>(
+        basedOn targets: GraphTargets,
+        testPlan: String?,
+        includedTargets: Set<String>,
+        excludedTargets: Set<String>,
+        excludingExternalTargets: Bool = false
+    ) -> Set<GraphTarget> where GraphTargets.Element == GraphTarget {
+        let allTestPlansTargetNames: Set<String>?
+        if includedTargets.isEmpty, let testPlanName = testPlan, let testPlan = self.testPlan(name: testPlanName) {
+            allTestPlansTargetNames = Set(testPlan.testTargets.filter(\.isEnabled).map(\.target.name))
+        } else {
+            allTestPlansTargetNames = nil
+        }
+
+        lazy var allInternalTargets = allInternalTargets().map(\.target.name)
+        return Set(
+            targets.filter { target in
+                if !includedTargets.isEmpty {
+                    return includedTargets.contains(target.target.name)
+                }
+                if excludedTargets.contains(target.target.name) {
+                    return false
+                }
+                if let allTestPlansTargetNames = allTestPlansTargetNames {
+                    return allTestPlansTargetNames.contains(target.target.name)
+                }
+                return excludingExternalTargets ? allInternalTargets.contains(target.target.name) : true
+            }
+        )
+    }
```

Comment:
Ah, nice catch! Indeed, the logic should be:

_1. If a test plan is specified, if the target target is not in the test plan, exclude it_
2. Include the target target if it's included in in includedTargets.
3. Exclude it if it's included in excludedTargets.
_4. If we're excluding external targets, only include the target if internal. If we're including external targets, include the target_.

Does that make more sense?

As for the name of the function, is it fine with that new logic, and otherwise what would you suggest? That's the best I had 😅